codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +72,1112 @@
+
+                do {
+                    Object m = inboundBuffer.poll();
+                    if (m == null) {
+                        break;
+                    }
+                    if (m == CLOSE_MESSAGE) {
+                        if (read) {
+                            pipeline().fireChannelReadComplete();
+                        }
+                        allocHandle.readComplete();
+                        DefaultHttp2StreamChannel.this.close();
+                        return;
+                    }
+                    doRead0((Http2Frame) m, allocHandle);
+                    read = true;
+                } while (allocHandle.continueReading());
+
+                allocHandle.readComplete();
+                pipeline().fireChannelReadComplete();
+            }
+
+            /**
+             * Returns whether reads should continue. The only reason reads shouldn't continue is that the
+             * channel was just closed.
+             */
+            void doRead0(Http2Frame frame, RecvByteBufAllocator.Handle allocHandle) {
+                int numBytesToBeConsumed = 0;
+                if (frame instanceof Http2DataFrame) {
+                    numBytesToBeConsumed = ((Http2DataFrame) frame).flowControlledBytes();
+                    allocHandle.lastBytesRead(numBytesToBeConsumed);
+                } else {
+                    if (frame instanceof Http2WindowUpdateFrame) {
+                        Http2WindowUpdateFrame windowUpdate = (Http2WindowUpdateFrame) frame;
+                        incrementOutboundFlowControlWindow(windowUpdate.windowSizeIncrement(), true);
+                    }
+                    allocHandle.lastBytesRead(ARBITRARY_MESSAGE_SIZE);
+                }
+                allocHandle.incMessagesRead(1);
+                pipeline().fireChannelRead(frame);
+
+                if (numBytesToBeConsumed != 0) {
+                    try {
+                        onBytesConsumed(ctx, stream, numBytesToBeConsumed);
+                    } catch (Http2Exception e) {
+                        pipeline().fireExceptionCaught(e);
+                    }
+                }
+            }
+
+            @Override
+            public void write(Object msg, final ChannelPromise promise) {
+                if (!isActive()) {
+                    ReferenceCountUtil.release(msg);
+                    promise.setFailure(CLOSED_CHANNEL_EXCEPTION);
+                    return;
+                }
+
+                // After this point its not possible to cancel a write anymore.
+                if (!promise.setUncancellable()) {
+                    ReferenceCountUtil.release(msg);
+                    return;
+                }
+
+                final int bytes = FlowControlledFrameSizeEstimator.HANDLE_INSTANCE.size(msg);
+
+                incrementOutboundFlowControlWindow(-bytes, true);
+
+                // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)